### DISCUSSION TITLE
Control propagation of observable sequence

### DISCUSSION THREAD

**Discussion by petroschalas** [bonsai-rx/bonsai#879] [Q&A]
Hello amazing people of Bonsai!

Thank you for all the work you 've been putting into this. I have a question regarding the PlaySource(Audio) node. This node is activated whenever it receives an observable sequence, i.e. a button press or anything else. In my workflow I need to be able to select between different sound files but I've found that I cannot change the audio file played by the PlaySource while the workflow is running. So I ended up putting as many PlaySource nodes as the different audio files I need. However, I would like to be able to trigger the different PlaySource nodes in a controlled manner, while the workflow is running. Is there a way to allow an observable sequence to propagate to these nodes, using an equality comparison ? I cannot use the Equality node, because even when there is no equality, the zero output still activates the PlaySource.

Thank you very much for your help

**Comment by bruno-f-cruz** [comment]
Hi,

Could you include your current workflow? I am not sure what kind of logic you are looking for, but if you want to play two sounds controlled by two distinct events, you could use something like the attached workflow.
Hope it helps!

[sound.zip](https://github.com/bonsai-rx/bonsai/files/8520757/sound.zip)

**Comment by petroschalas** [comment]
Dear Bruno,

thank you very much for your response. So my problem is that when using the PropertySource to update the source name of the PlaySource node, the actual audio file does not change. The value of the source does indeed change on the drop-down menu (on the right of the screen) of the PlaySource node, but the sound played does not. In order to change the audio file that the PlaySource node plays, I have to stop the protocol, change the value and start it again. For my purpose, I want to be able to change the audio sources while the protocol is running. Ideally, I want to use something like the "Where" function that I saw on a marble diagram of a Bonsai ppt. Where the observable sequence is propagated only when a specific value (or group of values) is received. Since even the transmission of a boolean false value will trigger the PlaySource, I have found no way to reliably control its activation. 

Button press is one way, but I would like its activation to be controlled without user input, for example to play the sound (activate PlaySource) when mouse enters ROI. Entering ROI could trigger broadcasting of a bool True or a numeric value >0.

Thank you again for your time
[sound_trigger_example.bonsai.zip](https://github.com/bonsai-rx/bonsai/files/8532020/sound_trigger_example.bonsai.zip)
.

**Reply by bruno-f-cruz** [reply]
Can you attach the full workflow? I am not able to identify the logic that you mentioned instanciates the 'PropertySource' logic in the last workflow you attached.

**Reply by petroschalas** [reply]
[online_trigger.bonsai.zip](https://github.com/bonsai-rx/bonsai/files/8552790/online_trigger.bonsai.zip)

**Reply by bruno-f-cruz** [reply]
I am sorry for keeping on insisting on this point, but I just dont see in the attached workflow anything related to: "So my problem is that when using the PropertySource to update the source name of the PlaySource node, the actual audio file does not change"...
Am I missing something?

**Reply by petroschalas** [reply]
Hello Bruno, I apologise for the confusion. This is a functional protocol, that I wrote to circumvent the problem that I mentioned. But I assume it would be easy to test the scenario I described, even with the example protocol you attached in your first reply. 

So I would like to know if that's an issue only I am facing and if yes, what should I do to fix it. Also please let me know if there is a better way to demonstrate to you that the audio file does not change, because I thought that attaching a workflow wouldn't really indicate anything. As I mentioned the source name is updated in the PlaySource node, but what is played by the speakers is still the source that the protocol was initiated with. And the actual audio file was not getting updated while the bonsai workflow was running.

My second issue that I faced is with Gain change using the UpdateListener node. As you can see on my attached workflow I use a PythonTransform node to change the Gain values. The Gain of the audio file is getting updated reliably while the workflow is running but not fast enough. When I change the Gain value from a positive integer to zero, there is a couple milliseconds where the audio file is being played with the previously chosen Gain. It then goes to zero, and no sound is played but there is still some sound making it through. Is there a way to fix this delay or is it inherent in how the UpdateListener node works ?

Lastly, is there a node in Bonsai that offers the "Where" functionality ? (please see attached image). I would like to have a convenient way to control activation of the PlaySource node, and since this node is activated whenever anything gets transmitted to it, I haven't found a reliable way to control it that does not rely on a button press. The reason I am interested in the "Where" functionality is because in that way I would filter the messsages getting sent to PlaySource and activate it only through specific messages.

Again, thank you very much for your time and I apologise again for any confusion.

![Screenshot 2022-04-26 at 11 26 55](https://user-images.githubusercontent.com/67964267/165268830-9831f8e0-716b-465c-a602-61744f769239.png)

**Comment by bruno-f-cruz** [comment]
I will try to address your answer point by point.

`But I assume it would be easy to test the scenario I described, even with the example protocol you attached in your first reply.
(...)As I mentioned the source name is updated in the PlaySource node, but what is played by the speakers is still the source that the protocol was initiated with. And the actual audio file was not getting updated while the bonsai workflow was running.`

The problem is that words dont easily translate into Bonsai code. It would be much easier and efficient to have the workflow you are trying to use and that reproduces the behavior you are describing. The code that i have sent before seems to achieve what you described and it toggles between the two sounds. Without checking your workflow I wont be able to identify the exact problem you are having.


`The Gain of the audio file is getting updated reliably while the workflow is running but not fast enough`

In your workflow, the camera is controlling the update of the gain, is this a valid reason for the delay?

`Lastly, is there a node in Bonsai that offers the "Where" functionality ? 
`

This would be a `Condition` node.

cheers,
B
